src/hotspot/share/gc/serial/serialHeap.cpp

```diff
@@ -786,6 +632,28 @@ void SerialHeap::scan_evacuated_objs(YoungGenScanClosure* young_cl,
   guarantee(young_gen()->promo_failure_scan_is_complete(), "Failed to finish scan");
 }
 
+void SerialHeap::collect_at_safepoint(bool full) {
+  assert(SafepointSynchronize::is_at_safepoint(), "precondition");
+  if (GCLocker::check_active_before_gc()) {
+    return;
+  }
+  collect_at_safepoint_no_gc_locker(full);
+}
+
+void SerialHeap::collect_at_safepoint_no_gc_locker(bool full) {
+  assert(!GCLocker::is_active(), "precondition");
+  bool clear_soft_refs = must_clear_all_soft_refs();
+
+  if (!full) {
+    bool success = do_young_gc(clear_soft_refs);
+    if (success) {
+      return;
+    }
+    // Upgrade to Full-GC if young-gc fails
+  }
+  do_full_collection_no_gc_locker(clear_soft_refs);
+}
+
```

Comment:
Maybe the method `do_young_gc` can be renamed to `do_young_collection` or `do_young_collection_no_gc_locker` which is consistent with `do_full_collection` or `do_full_collection_no_gc_locker`.